File: Codechef/gcd.py

def gcd(a,b):
	if b==0:
		return a
	return gcd(b,a%b)

def repeat(time,p,t):
	r=1
	while time > 0:
		if(time%2) :
			r = (r*p)%t
		p = (p*p)%t
		time = time//2
	return r%t



q = int(input())
prime = 1000000007
i = 0
while i<q:
	a,b,n = raw_input().split(" ")
	n = int(n)
	b = int(b)
	a = int(a)
	mod = a-b
	if mod == 0:
		an = repeat(n,a%prime,prime)
		ans = (an + an)%prime
		print(ans)
	else :	
		an = repeat(n,a%mod,mod)
		bn = repeat(n,b%mod,mod)
		ans = gcd(mod,(an+bn)%mod)%prime
		print(ans)
	i+=1

File: Codechef/gcd.cpp
#include <bits/stdc++.h>

#define PRIME 1000000007
#define ul int_128

using namespace std;

ul gcd(ul a, ul b)
{
	if(b==0)
	return a;
	return gcd(b,a%b);
}
ul prd(ul a,ul b,ul t)
{
	ul r = 0,y=a;
	while(b>0)
	{
		if(b&1)
		r = (r+y)%t;
		y = (y + y)%t;
		b >>= 1;
	}
	return r;
}
ul repeat(ul time,ul p,ul t)
{
	ul r=1;
	while(time)
	{
		if(time%2)
		r = prd(r,p,t);
		p = prd(p,p,t);
		time/=2;
	}
	return r;
}

int main()
{
	int q;
	scanf("%d",&q);
	while(q--)
	{
		ul a,b,n,mod,ans,an,bn;
		//scanf("%lu%lu%lu",&a,&b,&n);
		cin>>a>>b>>n
		mod = abs(a-b);
		if(mod == 0)
		{
			an = repeat(n,a%PRIME,PRIME);
			bn = repeat(n,b%PRIME,PRIME);
			ans = (an + bn)%PRIME;
			printf("%lu\n",ans);
			continue;
		}
		an = repeat(n,a%mod,mod);
		bn = repeat(n,b%mod,mod);
		ans = gcd(mod,(an+bn)%mod)%PRIME;
		cout<<ans<<endl;
		//printf("%lu\n",ans);
		
	}
	return 0;
}
